### PR TITLE
chore:core: add tests for todos in prepo market

### PR DESCRIPTION
**Pull request by nmimran99**
chore:core: add tests for todos in prepo market

Add 2 tests regarding different fee levels of redeem hook and the amount of collateral the redeemer is getting based on the fee taken.

The PR adds a `TestRedeemHook.sol` contract that acts as a hook that keeps 50% of the fees to make the test possible

### CODE REVIEW

File: apps/smart-contracts/core/contracts/test/TestRedeemHook.sol [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+import "../interfaces/IHook.sol";
+import "../interfaces/IPrePOMarket.sol";
+import "prepo-shared-contracts/contracts/AccountListCaller.sol";
+import "prepo-shared-contracts/contracts/AllowedMsgSenders.sol";
+import "prepo-shared-contracts/contracts/SafeOwnable.sol";
+import "prepo-shared-contracts/contracts/TokenSenderCaller.sol";
+import "prepo-shared-contracts/contracts/TreasuryCaller.sol";
+
+contract TestRedeemHook is
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/contracts/test/TestRedeemHook.sol:12]: Ok I think you made this unnecessarily complicated. This contract only needs to inherit `IHook`, and implement a `hook`. The hook can just `transferFrom(msg.sender, address(this), fee - 1)`.

File: apps/smart-contracts/core/test/unit/PrePOMarket.test.ts [hunk 2]
```diff
@@ -929,8 +934,67 @@ describe('=> prePOMarket', () => {
       expect(await collateralToken.balanceOf(user.address)).to.eq(totalOwed)
     })
 
-    // TODO ADD 'sends correct collateral amount if hook takes partial fee'
-    // TODO ADD 'sends correct collateral amount if hook takes full fee'
+    it('sends correct collateral amount if hook takes partial fee', async () => {
+      // TestRedeemHook meant to take 50% of the fee
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/unit/PrePOMarket.test.ts:938]: So given that you simplify the `TestRedeemHook` contract you created, you don't need to reinitialize all these things. You can just set the `redeemHook` of the `prePOMarket` already deployed to the test redeem hook for the purposes of the test, and set it back to the actual one later if need be.

**Review by ramenforbreakfast — CHANGES_REQUESTED** [review]

**Review by ramenforbreakfast — APPROVED** [review]
LGTM, can you remove TODOs as well? I think we can forego the full fee transfer testcase.

File: apps/smart-contracts/core/contracts/test/TestRedeemHook.sol [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity =0.8.7;
+
+import "../interfaces/IHook.sol";
+import "../interfaces/IPrePOMarket.sol";
+
+contract TestRedeemHook is IHook {
+  /// @notice this contract is a test contract made to test a use case where the redeem hook only takes a part of the fee
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/test/TestRedeemHook.sol:8]: nit: move comment above the `contract TestRedeemHook` line

File: apps/smart-contracts/core/test/unit/PrePOMarket.test.ts [hunk 4]
```diff
@@ -929,8 +934,54 @@ describe('=> prePOMarket', () => {
...
+      await prePOMarket.connect(treasury).setRedeemHook(testRedeemHook.address)
+      const totalOwed = await calculateTotalOwed(longToRedeem, shortToRedeem, false)
+      expect(totalOwed).to.be.gt(0)
+      const redeemFee = calculateFee(totalOwed, await prePOMarket.getRedemptionFee())
+      expect(redeemFee).to.be.gt(0)
+      const expectedPartialFee = redeemFee.div(2)
+
+      await prePOMarket.connect(user).redeem(longToRedeem, shortToRedeem, user.address)
+
+      expect(await collateralToken.balanceOf(user.address)).to.eq(totalOwed.sub(expectedPartialFee))
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/unit/PrePOMarket.test.ts:953]: also check the `Redemption` event emitted here

File: apps/smart-contracts/core/test/unit/PrePOMarket.test.ts [hunk 5]
```diff
@@ -929,8 +934,54 @@ describe('=> prePOMarket', () => {
...
+      const totalOwed = await calculateTotalOwed(longToRedeem, shortToRedeem, false)
+      expect(totalOwed).to.be.gt(0)
+      const redeemFee = calculateFee(totalOwed, await prePOMarket.getRedemptionFee())
+      expect(redeemFee).to.be.gt(0)
+      await prePOMarket.connect(user).redeem(longToRedeem, shortToRedeem, user.address)
+
+      expect(await collateralToken.balanceOf(user.address)).eq(totalOwed.sub(redeemFee))
+      expect(await collateralToken.balanceOf(treasury.address)).eq(
+        treasuryCollateralBefore.add(redeemFee)
+      )
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/unit/PrePOMarket.test.ts:983]: also check the `Redemption` event emitted here

File: apps/smart-contracts/core/test/unit/PrePOMarket.test.ts [hunk 6]
```diff
@@ -929,8 +934,54 @@ describe('=> prePOMarket', () => {
...
+      await testRedeemHook.connect(deployer).setTokenSender(tokenSender.address)
+      await prePOMarket.connect(treasury).setRedeemHook(testRedeemHook.address)
+      allowedCallers.isIncluded.whenCalledWith(prePOMarket.address).returns(true)
+      allowList.isIncluded.whenCalledWith(user.address).returns(true)
+
+      const totalOwed = await calculateTotalOwed(longToRedeem, shortToRedeem, false)
+      expect(totalOwed).to.be.gt(0)
+      const redeemFee = calculateFee(totalOwed, await prePOMarket.getRedemptionFee())
+      expect(redeemFee).to.be.gt(0)
+      await prePOMarket.connect(user).redeem(longToRedeem, shortToRedeem, user.address)
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/unit/PrePOMarket.test.ts:978]:
```suggestion
      const totalOwed = await calculateTotalOwed(longToRedeem, shortToRedeem, false)
      expect(totalOwed).to.be.gt(0)
      const redeemFee = calculateFee(totalOwed, await prePOMarket.getRedemptionFee())
      expect(redeemFee).to.be.gt(0)
      
      await prePOMarket.connect(user).redeem(longToRedeem, shortToRedeem, user.address)
```

File: apps/smart-contracts/core/test/unit/PrePOMarket.test.ts (hunk 4 above)
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/unit/PrePOMarket.test.ts:953]: also check treasury collateral balance change (like in test below)

**Review by XavierEkkel — APPROVED** [review]
LGTM

**Pull request merged**